glass-easel/src/shadow_root.ts

```diff
@@ -399,307 +398,333 @@ export class ShadowRoot extends VirtualNode {
-      }
-      this._$host.childNodes.every(rec)
+  forEachNodeInSlot(f: (node: Node, slot: Element | null | undefined) => boolean | void) {
+    const childNodes = this._$host.childNodes
+    for (let i = 0; i < childNodes.length; i += 1) {
+      if (!Element.forEachNodeInSlot(childNodes[i]!, f)) return false
     }
+    return true
   }
 
   /**
-   * Iterate elements in specified slot
+   * Iterate elements in specified slot (slots-inherited nodes included)
    */
-  forEachNodeInSpecifiedSlot(slot: Element | string, f: (node: Node, posIndex: number) => void) {
-    if (this._$singleSlot) {
-      if (slot !== '' && this._$singleSlot !== slot) return
-      const rec = (child: Node, index: number) => {
-        f(child, index)
-        if (child instanceof Element) {
-          if (child._$inheritSlots) {
-            child.childNodes.forEach(rec)
-          }
-        }
+  forEachNodeInSpecifiedSlot(slot: Element | null, f: (node: Node) => boolean | void): boolean {
+    if (slot) {
+      const childNodes = slot.slotNodes!
+      for (let i = 0; i < childNodes.length; i += 1) {
+        const node = childNodes[i]!
+        if (f(node) === false) return false
       }
-      this._$host.childNodes.forEach(rec)
-    } else if (this._$dynamicSlotSet) {
-      if (typeof slot === 'string') return
-      const recInheritedSlot = (child: Node, index: number) => {
-        f(child, index)
-        if (child instanceof Element && child._$inheritSlots) {
-          child.childNodes.forEach(recInheritedSlot)
-        }
+      return true
+    }
+    return this.forEachNodeInSlot((node, slot) => (slot === null ? f(node) : true))
+  }
+
+  /**
+   * Iterate elements with their slots (slots-inherited nodes not included)
+   */
+  forEachSlotContentInSlot(f: (node: Node, slot: Element | null | undefined) => boolean | void) {
+    const childNodes = this._$host.childNodes
+    for (let i = 0; i < childNodes.length; i += 1) {
+      if (!Element.forEachSlotContentInSlot(childNodes[i]!, f)) return false
+    }
+    return true
+  }
+
+  /**
+   * Iterate elements in specified slot (slots-inherited nodes not included)
+   */
+  forEachSlotContentInSpecifiedSlot(
+    slot: Element | null,
+    f: (node: Node) => boolean | void,
+  ): boolean {
+    if (slot) {
+      const childNodes = slot.slotNodes!
+      for (let i = 0; i < childNodes.length; i += 1) {
+        const node = childNodes[i]!
+        if (!node._$inheritSlots && f(node) === false) return false
       }
-      const rec = (child: Node, index: number) => {
-        if (this.getContainingSlot(child) !== slot) return
-        f(child, index)
-        if (child instanceof Element && child._$inheritSlots) {
-          child.childNodes.forEach(recInheritedSlot)
+      return true
+    }
+    return this.forEachSlotContentInSlot((node, slot) => (slot === null ? f(node) : true))
+  }
+
+  /**
+   * Check whether a node is attached to this shadow root
+   */
+  isAttached(node: Node): boolean {
```

Comment:
Not a good name. This is not the same as the attach status.

Consider `connectedToShadowRoot` ?

Reply:
> Not a good name. This is not the same as the attach status.
> 
> Consider `connectedToShadowRoot` ?

Renamed to `isConnected`